watson_developer_cloud/websocket/recognize_listener.py

```diff
@@ -0,0 +1,199 @@
+            transcripts.append(transcript)
+        return transcripts
+
+    def send(self, data, opcode=websocket.ABNF.OPCODE_TEXT):
+        """
+        Send message to server.
+
+        data: message to send. If you set opcode to OPCODE_TEXT,
+              data must be utf-8 string or unicode.
+        opcode: operation code of data. default is OPCODE_TEXT.
+        """
+        self.ws_client.send(data, opcode)
+
+    def send_audio(self, ws):
+        """
+        Stream audio to server
+
+        :param ws: Websocket client
+        """
+        def run(*args):
+            """Background process to stream the data"""
+            if not self.audio_source.is_buffer:
+                while True:
+                    chunk = self.audio_source.input.read(ONE_KB)
+                    if not chunk:
+                        break
+                    self.ws_client.send(chunk, websocket.ABNF.OPCODE_BINARY)
+                    time.sleep(TEN_MILLISECONDS)
+
+                self.audio_source.input.close()
+            else:
+                while True:
+                    try:
+                        if not self.audio_source.input.empty():
+                            chunk = self.audio_source.input.get()
+                            self.ws_client.send(chunk, websocket.ABNF.OPCODE_BINARY)
+                            time.sleep(TEN_MILLISECONDS)
+                        if self.audio_source.input.empty():
+                            if self.audio_source.is_recording:
+                                time.sleep(TEN_MILLISECONDS)
+                            else:
+                                break
+                    except Exception:
+                        if self.audio_source.is_recording:
+                            time.sleep(TEN_MILLISECONDS)
+                        else:
+                            break
+
+            time.sleep(TEN_MILLISECONDS)
+            self.ws_client.send(self.build_closing_message(), websocket.ABNF.OPCODE_TEXT)
+            ws.close()
+
+        thread.start_new_thread(run, ())
+
+    def on_open(self, ws):
+        """
+        Callback executed when a connection is opened to the server.
+        Handles streaming of audio to the server.
+
+        :param ws: Websocket client
+        """
+        self.callback.on_connected()
+
+        # Send initialization message
+        init_data = self.build_start_message(self.options)
+        self.ws_client.send(json.dumps(init_data).encode('utf8'), websocket.ABNF.OPCODE_TEXT)
+
+    def on_data(self, ws, message, message_type, fin):
+        """
+        Callback executed when message is received from the server.
+
+        :param ws: Websocket client
+        :param message: utf-8 string which we get from the server.
+        :param message_type: Message type which is either ABNF.OPCODE_TEXT or ABNF.OPCODE_BINARY
+        :param fin: continue flag. If 0, the data continues.
+        """
+        json_object = json.loads(message)
```

Comment:
Sometimes the data can't be parsed. You need to take that into account. Maybe a try catch